src/Utils.php

```diff
@@ -193,7 +193,7 @@ static function ($a, $b) {
      */
     public static function naturalLanguageJoinWithBackticks(array $names)
     {
-        if (empty($names)) {
+        if ([] === $names) {
```

Comment:
`$names` is guaranteed to be declared. There's no reason to use empty here. This line does not check if input is array, but array with no elements.

Reply:
> $names is guaranteed to be declared
Correct, that's why there's no need for `isset`. The check is for empty though.

What do you ask if you have an array, and want to know if it's empty?
"Is this array empty?"
You do not ask
"Is this variable strictly equal to the empty array"

Therefor, both DX-wise and syntactically `if (empty($names))` is more correct. It conveys the meaning more clear.

Reply:
`empty` does `isset`. Like most people using it for "semantical" purposes, you don't understand it doesn't do what it's name says it does. It's freaking misleading, don't use it! It does same thing following does:
```php 
!isset($var) || !$var)
```

Its "awesome" semanticism is especially shown when using collection objects instead of arrray like

```php
$collection = new ArrayObject([]);

var_dump(empty($collection));
```

Reply:
> `empty` does `isset`.

I don't care that under the hood it **also** does isset, that is besides the point.

> Like most people using it for "semantical" purposes, you don't understand it doesn't do what it's name says it does.

Don't belittle me be saying I don't know something when in fact you do not know if I know or not. There's no need to attack people.

> Its "awesome" semanticism is especially shown when using collection objects instead of arrray like
> 
> ```
> $collection = new ArrayObject([]);
> 
> var_dump(empty($collection));
> ```

What do ArrayObjects matter when we're guaranteed an array? Your entire argument in this post is totally missing the premise under which I make mine.

Reply:
> I don't care that under the hood it also does isset, that is besides the point.

It's not besides the point, because you really don't want to do isset for declared variables, even if it's under the hood. Code will silently keep working when variable will no longer be declared instead of telling you it's undeclared. This is undesired. Use some empty helper which don't have that ugly behaviour. 

> What do ArrayObjects matter when we're guaranteed an array? 

Exactly, code is guaranteed to be an array, so we can use `[] === $names` instead of `empty` which don't care what it processes, swallows literally anything, is misleading and does extra things we don't want. ArrayObjects matter as a proof that empty doesn't check for emptiness. It doesn't do what it says it does, why would you use such misleading function? Do you like to lie to yourself?

Reply:
> Do you like to lie to yourself?

How about next time you just leave that sentence out?

Look, I get that you have a dislike for empty, fine. All I was saying was _at the current line of code_, empty is fine, and it does exactly what you want from it, and that has been my point the entire time.

You keep saying _"no empty is not fine, because [generic reason empty is not fine]"_. Yes I see that there are cases where empty is or could be bad. However, something can both have good and bad applications.

Don't just throw away the good with the bad.

Reply:
For me it is similar to using static methods - yes, they are perfectly correct construction - but, yet we do not recommend it. With an empty I saw many people are abusing it. We could say don't use it when it's not OK, but we all know how it ends.

On the other side if we decide to try not to use it at all - maybe also as a training - and think about it as of 10th [calisthenic](https://williamdurand.fr/2013/06/03/object-calisthenics) we could learn something.